### PR TITLE
Feature/sg 858 ignore multiple labels segmentation metrics support

### PR DESCRIPTION
**Pull request by shaydeci**
Added the option for multiple ignored indices for our segmentation metrics + unit tests.

### CODE REVIEW

File: src/super_gradients/training/metrics/segmentation_metrics.py [hunk 1]
```diff
@@ -122,6 +122,18 @@ def intersection_and_union(im_pred, im_lab, num_class):
     return area_inter, area_union
 
 
+def _map_ignored_inds(target: torch.Tensor, ignore_index_list, unfiltered_num_classes) -> torch.Tensor:
```
Comment (BloodAxe) [inline src/super_gradients/training/metrics/segmentation_metrics.py:125]: I feel this method needs docs, especially what is unfiltered_num_classes.
Maybe think of a better name for this variable? `num_evaluated_classes` maybe?

File: src/super_gradients/training/metrics/segmentation_metrics.py [hunk 2]
```diff
@@ -174,27 +205,73 @@ def __init__(self, ignore_label=-100, dist_sync_on_step=False, metrics_args_prep
...
+
     def compute(self):
         _total_correct = self.total_correct.cpu().detach().numpy().astype("int64")
         _total_label = self.total_label.cpu().detach().numpy().astype("int64")
         pix_acc = np.float64(1.0) * _total_correct / (np.spacing(1, dtype=np.float64) + _total_label)
         return pix_acc
 
 
+def _handle_multiple_ignored_inds(ignore_index, num_classes):
+    if isinstance(ignore_index, list):
```
Comment (BloodAxe) [inline src/super_gradients/training/metrics/segmentation_metrics.py:236]: We have many similar checks `if is list do that else do this` with respect to ignore_index.
Maybe it is possible to somehow refactor to have this check in one function and then reuse it in metrics?

File: src/super_gradients/training/metrics/segmentation_metrics.py [hunk 3]
```diff
@@ -227,12 +333,19 @@ def __init__(
...
+
+        self.ignore_index_list = ignore_index_list
+        self.unfiltered_num_classes = unfiltered_num_classes
         self.metrics_args_prep_fn = metrics_args_prep_fn or PreprocessSegmentationMetricsArgs(apply_arg_max=True)
         self.greater_is_better = True
 
     def update(self, preds, target: torch.Tensor):
         preds, target = self.metrics_args_prep_fn(preds, target)
+        if self.ignore_index_list is not None:
+            target = _map_ignored_inds(target, self.ignore_index_list, self.ignore_index)
```
Comment (BloodAxe) [inline src/super_gradients/training/metrics/segmentation_metrics.py:348]: Dice metric remaps predictions. This function doesn't. Is it intended?

File: src/super_gradients/training/metrics/segmentation_metrics.py [hunk 4]
```diff
@@ -174,27 +205,73 @@ def __init__(self, ignore_label=-100, dist_sync_on_step=False, metrics_args_prep
...
+        labeled_mask = self._handle_multiple_ignored_inds(target)
 
-        labeled_mask = target.ne(self.ignore_label)
         pixel_labeled = torch.sum(labeled_mask)
         pixel_correct = torch.sum((predict == target) * labeled_mask)
         self.total_correct += pixel_correct
         self.total_label += pixel_labeled
 
+    def _handle_multiple_ignored_inds(self, target):
+        if isinstance(self.ignore_label, list):
```
Comment (BloodAxe) [inline src/super_gradients/training/metrics/segmentation_metrics.py:216]: Just a slight optimization, which may be faster:
labeled_mask = torch.ones_like(target)
for ignored_label:
   labeled_mask.masked_fill(target.eq(ignored_label), 0)

This avoids creating temporary buffers and should be a bit faster.

File: src/super_gradients/training/metrics/segmentation_metrics.py [hunk 5]
```diff
@@ -174,27 +205,73 @@ def __init__(self, ignore_label=-100, dist_sync_on_step=False, metrics_args_prep
...
+            labeled_mask = None
+            for ignored_label in self.ignore_label:
+                if labeled_mask is None:
+                    labeled_mask = target.ne(ignored_label)
+                else:
+                    labeled_mask = torch.logical_and(labeled_mask, target.ne(ignored_label))
+        else:
+            labeled_mask = target.ne(self.ignore_label)
+
+        return labeled_mask
```
Comment (BloodAxe) [inline src/super_gradients/training/metrics/segmentation_metrics.py:226]: Maybe name if `evaluated_classes_mask` ?

File: src/super_gradients/training/metrics/segmentation_metrics.py (hunk 1 above)
Comment (shaydeci) [reply src/super_gradients/training/metrics/segmentation_metrics.py]: Added c681c262f3e962b1faa3317e7fca7dd7d7bdbb37

File: src/super_gradients/training/metrics/segmentation_metrics.py [hunk 6]
```diff
@@ -174,27 +205,73 @@ def __init__(self, ignore_label=-100, dist_sync_on_step=False, metrics_args_prep
...
+
     def compute(self):
         _total_correct = self.total_correct.cpu().detach().numpy().astype("int64")
         _total_label = self.total_label.cpu().detach().numpy().astype("int64")
         pix_acc = np.float64(1.0) * _total_correct / (np.spacing(1, dtype=np.float64) + _total_label)
         return pix_acc
 
 
+def _handle_multiple_ignored_inds(ignore_index, num_classes):
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/metrics/segmentation_metrics.py]: Added c681c262f3e962b1faa3317e7fca7dd7d7bdbb37

File: src/super_gradients/training/metrics/segmentation_metrics.py (hunk 3 above)
Comment (shaydeci) [reply src/super_gradients/training/metrics/segmentation_metrics.py]: Nice catch!
Fixed

File: src/super_gradients/training/metrics/segmentation_metrics.py (hunk 5 above)
Comment (shaydeci) [reply src/super_gradients/training/metrics/segmentation_metrics.py]: sure, done in a9351123c4a0f04b854f552cf04291dacec9500f

File: src/super_gradients/training/metrics/segmentation_metrics.py (hunk 4 above)
Comment (shaydeci) [reply src/super_gradients/training/metrics/segmentation_metrics.py:233]: might be missing something, but don't see why it would faster (since masked_fill is not in place).
Nonetheless I updated it.

File: src/super_gradients/training/metrics/segmentation_metrics.py [hunk 7]
```diff
@@ -174,46 +222,136 @@ def __init__(self, ignore_label=-100, dist_sync_on_step=False, metrics_args_prep
...
+      ignore_index=0,
+    And since we map all of the ignored indices to 0, it is if we removed them and introduces a new index:
+      num_classes = num_classes - len(ignore_index_list) +1
+    Unfiltered num_classes is used in .update() for mapping of the original indice values.
+    Sets ignore_index to 0
+    :param ignore_index: list or single int representing the class ind(ices) to ignore.
+    :param num_classes: int, num_classes (original, before mapping) being passed to segmentation metric classesץ
+    :return:ignore_index, ignore_index_list, num_classes, unfiltered_num_classesignore_index, ignore_index_list, num_classes, unfiltered_num_classes
+    """
+    if isinstance(ignore_index, list):
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/metrics/segmentation_metrics.py]: good suggestion, fixed in latest commit.

File: src/super_gradients/training/metrics/segmentation_metrics.py [hunk 8]
```diff
@@ -174,46 +222,136 @@ def __init__(self, ignore_label=-100, dist_sync_on_step=False, metrics_args_prep
...
-        ignore_index: Optional[int] = None,
+        ignore_index: Optional[Union[int, List[int]]] = None,
         reduction: str = "elementwise_mean",
         threshold: float = 0.5,
         metrics_args_prep_fn: Optional[AbstractMetricsArgsPrepFn] = None,
     ):
 
         if num_classes <= 1:
             raise ValueError(f"IoU class only for multi-class usage! For binary usage, please call {BinaryIOU.__name__}")
+        if isinstance(ignore_index, list) and reduction == "none":
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/metrics/segmentation_metrics.py]: fixed in latest commit as well

**Review by BloodAxe — APPROVED** [review]
LGTM
